modules/perSystem.nix

```diff
@@ -59,14 +59,21 @@ let
 in
 {
   options = {
-    systems = mkOption {
+    systems = mkOption ({
       description = ''
         All the system types to enumerate in the flake output subattributes.
 
         In other words, all valid values for `system` in e.g. `packages.<system>.foo`.
+
+        If the `systems` flake input exists, it will be imported as the
+        default. See <https://github.com/nix-systems/nix-systems> for a
+        description of the pattern.
       '';
       type = types.listOf types.str;
-    };
+    } // (lib.optionalAttrs self.inputs?systems) {
+      # Load the systems input by default if it has been declared.
+      default = import self.inputs.systems;
+    });
```

Comment:
This doesn't really fit the architecture, because we shouldn't be making assumptions about the user's flake inputs; certainly not in the core, non-optional modules.

It'd be nicer to have nix-systems implement the flake-parts interface, by providing a module that just sets the `systems` option. In that case it doesn't need to be a default either.

Reply:
Does flake-parts auto-import inputs' flakeModules?

Alternatively, I would propose extending the flake-parts "systems" type to support `inherits systems;` when systems is an input.

Reply:
By default it does try to import all inputs, and I wouldn't encourage that, because that would defeat lazy fetching and any solution to #119 
Accepting a flake (I presume?) for an arbitrary option does not seem like a sensible solution to me. It abbreviates the syntax, but that is not a goal. An option should have a meaningful type.

Reply:
> This doesn't really fit the architecture, because we shouldn't be making assumptions about the user's flake inputs; certainly not in the core, non-optional modules.

This already happens with the Nixpkgs module's `_module.args.pkgs` assuming `inputs.nixpkgs`. This is justified through it being an important, special case. I do think the same could be said here.

> It'd be nicer to have nix-systems implement the flake-parts interface, by providing a module that just sets the `systems` option. In that case it doesn't need to be a default either.

`nix-systems/default` *can't* provide a module. All consumers have to assume that `inputs.systems.flake = false;` for the simple overrides this pattern allows to work properly. (`inputs.systems.url = "file:./systems.nix";` is very intentionally valid.)

I would recommend changing the fancy conditional attrset merge here with normal
```nix
mkOption {
  # …
  default = import inputs.systems or (throw "flake-parts: The flake does not have a `systems` input. Please add it, or set `systems` yourself.");
  defaultText = literalExpression ''import inputs.systems'';
}
```
like the Nixpkgs module does.

Reply:
> This already happens with the Nixpkgs module's `_module.args.pkgs` assuming `inputs.nixpkgs`.

That's a temporary measure and a bad precedent, and documented as such in the code. I will remove it.

> `"file:./systems.nix"`

I wasn't aware of that requirement.

I think the standard way is good enough then. All this fancy `system` inheriting stuff is just a ridiculous workaround anyway.

```
systems = import inputs.systems;
```

Reply:
> I think the standard way is good enough then. All this fancy `system` inheriting stuff is just a ridiculous workaround anyway.
> 
> ```
> systems = import inputs.systems;
> ```

That's fair; the idea behind this would be to reduce boilerplate just a little bit more and push people towards just using `inputs.systems` instead of deciding to list the inputs out inline.

Reply:
> > This already happens with the Nixpkgs module's `_module.args.pkgs` assuming `inputs.nixpkgs`.
> 
> That's a temporary measure and a bad precedent, and documented as such in the code. I will remove it.

It's probably better to discuss this more in <https://github.com/hercules-ci/flake-parts/pull/137>, but I think it would be a shame to see that module go before a capable `inputs.nixpkgs.flakeModules.default` exists. Virtually every flake will need Nixpkgs, and having to manually set up `config._module.args.pkgs` boilerplate or import a 3rd-party flake module for it will push people away from using flake-parts when they're first evaluating it. At the very least, some sort of `inputs.nixpkgs-flake-parts.url = "github:hercules-ci/nixpkgs-flake-parts";` needs to exist first, before this is removed. I think the Nixpkgs module is an okay exception until flake-parts picks up more momentum.

Reply:
@roberth what is your conclusion? I wouldn't mind re-opening the PR if it's possible to move forwards.